File: hostel_management_system_backend/Controllers/HostelListingsController.cs
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public sealed class HostelListingsController : ControllerBase
{
    private readonly IHostelListingsService _service;

    public HostelListingsController(IHostelListingsService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<ActionResult<List<HostelListingReadDto>>> GetAll(CancellationToken cancellationToken)
    {
        var listings = await _service.GetAllAsync(cancellationToken);
        return Ok(listings);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<HostelListingReadDto>> GetById(Guid id, CancellationToken cancellationToken)
    {
        var listing = await _service.GetByIdAsync(id, cancellationToken);
        return listing is null ? NotFound() : Ok(listing);
    }

    [HttpPost]
    public async Task<ActionResult<HostelListingReadDto>> Create([FromBody] HostelListingCreateDto dto, CancellationToken cancellationToken)
    {
        var created = await _service.CreateAsync(dto, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<HostelListingReadDto>> Update(Guid id, [FromBody] HostelListingUpdateDto dto, CancellationToken cancellationToken)
    {
        var updated = await _service.UpdateAsync(id, dto, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken cancellationToken)
    {
        var deleted = await _service.DeleteAsync(id, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: src/hostels/listings/hostel-listings.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';

import {
  HostelListingCreateDto,
  HostelListingReadDto,
  HostelListingUpdateDto,
} from './dto/hostel-listings.dto';
import { HostelListingsService } from './hostel-listings.service';

@Controller('hostellistings')
export class HostelListingsController {
  constructor(private readonly hostelListingsService: HostelListingsService) {}

  @Get()
  getAll(): Promise<HostelListingReadDto[]> {
    return this.hostelListingsService.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseUUIDPipe) id: string): Promise<HostelListingReadDto> {
    return this.hostelListingsService.getById(id);
  }

  @Post()
  create(@Body() dto: HostelListingCreateDto): Promise<HostelListingReadDto> {
    return this.hostelListingsService.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: HostelListingUpdateDto,
  ): Promise<HostelListingReadDto> {
    return this.hostelListingsService.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseUUIDPipe) id: string) {
    await this.hostelListingsService.delete(id);
  }
}
